### ISSUE TITLE
GPU训练显存分配失败问题

### ISSUE THREAD

**Issue by lxwzafu** [PaddlePaddle/Paddle#20084]
- 版本、环境信息：
    1）PaddlePaddle版本：paddlepaddle-1.5.2
    2）GPU：训练用GPU型号： GeForce GTX 1080 Ti，驱动版本25.21.14.1896    这样的GPU共2块
    3）CUDA 10.0.130_411.31， CUDNN 10.0 v7.5.0.56
    4）系统环境：Windows10 64位、VS2017、Python36_64
- 训练信息
    1）单机，多卡 （但是目前训练好像只用到了一张卡）
    2）显存信息：每张显卡：专用GPU内存 11GB GDDR5X，共享GPU内存 31.8GB，GPU内存 42.8GB
    3）SSD模型训练，链接：https://aistudio.baidu.com/aistudio/projectdetail/78972
在控制台显示以下信息：
start ssd, train params: {'input_size': [3, 300, 300], 'class_dim': 2, 'label_dict': {'background': 0, 'person': 1}, 'image_count': 7104, 'log_feed_image': False, 'pretrained': True, 'pretrained_model_dir': './pretrained-model', 'continue_train': True, 'save_model_dir': './ssd-model', 'model_prefix': 'mobilenet-ssd', 'data_dir': 'Y:/pascalvoc', 'mean_rgb': [127.5, 127.5, 127.5], 'file_list': 'train_person.txt', 'eval_file_list': 'eval_person.txt', 'label_list': 'label_list_person', 'mode': 'train', 'num_epochs': 400, 'train_batch_size': 64, 'use_gpu': True, 'apply_distort': True, 'apply_expand': True, 'apply_corp': True, 'image_distort_strategy': {'expand_prob': 0.5, 'expand_max_ratio': 4, 'hue_prob': 0.5, 'hue_delta': 18, 'contrast_prob': 0.5, 'contrast_delta': 0.5, 'saturation_prob': 0.5, 'saturation_delta': 0.5, 'brightness_prob': 0.5, 'brightness_delta': 0.125}, 'rsm_strategy': {'learning_rate': 0.001, 'lr_epochs': [40, 60, 80, 100], 'lr_decay': [1, 0.5, 0.25, 0.1, 0.01]}, 'momentum_strategy': {'learning_rate': 0.1, 'decay_steps': 128, 'decay_rate': 0.8}, 'early_stop': {'sample_frequency': 50, 'successive_limit': 3, 'min_loss': 1.28, 'min_curr_map': 0.86}}
W0929 14:21:07.261186 35732 device_context.cc:259] Please NOTE: device: 0, CUDA Capability: 61, Driver API Version: 10.1, Runtime API Version: 10.0
W0929 14:21:07.275125 35732 device_context.cc:267] device: 0, cuDNN Version: 7.5.
load param from retrain model
current round: 0, start read image
W0929 14:21:15.004384 35732 system_allocator.cc:121] Cannot malloc 45.1252 MB GPU memory. Please shrink FLAGS_fraction_of_gpu_memory_to_use or FLAGS_initial_gpu_memory_in_mb or FLAGS_reallocate_gpu_memory_in_mbenvironment variable to a lower value. Current FLAGS_fraction_of_gpu_memory_to_use value is 0.5. Current FLAGS_initial_gpu_memory_in_mb value is 0. Current FLAGS_reallocate_gpu_memory_in_mb value is 0
F0929 14:21:15.005416 35732 legacy_allocator.cc:201] Cannot allocate 45.125000MB in GPU 0, available 6.637500MBtotal 11811160064GpuMinChunkSize 256.000000BGpuMaxChunkSize 4.278862GBGPU memory used: 8.760785GB
*** Check failure stack trace: ***

然后跳出：调试适配器已意外退出，程序退出运行状态

把train_batch_size从64调整到32能正常运行。但是有这么大的显存，为什么这么点显存分配还会失败？可以如何设置使用多个GPU和显存管理策略来提高训练效率？

**Comment by Exception-star** [comment]
只有11G，用了8.7G，可用6M，无法分配45M

**Comment by lxwzafu** [comment]
意思是，这块GPU配置过低，出现这种情况是正常的是吗？机器没有在运行其他程序，GPU基本上就是这个程序在用。
还有，程序会无规律的出现死机状态，就是程序不使用CPU和GPU了，内存和显存不进行数据交换（数值不发生变化）但是保持原占用状态。这种情况碰到了2次，一次发生在重复训练了3轮后的某一阶段，另一次发生在第二次重复训练的某一阶段。但是扔在那里不管他，又会莫名其妙继续，然后又可能卡在那里（第二次就是这样）。但是我最长重复训练了34轮也没碰到这种情况。不知是怎么回事？以下是在训练过程中控制台打印的内容：
start ssd, train params: {'input_size': [3, 300, 300], 'class_dim': 2, 'label_dict': {'background': 0, 'person': 1}, 'image_count': 7104, 'log_feed_image': False, 'pretrained': True, 'pretrained_model_dir': './pretrained-model', 'continue_train': True, 'save_model_dir': './ssd-model', 'model_prefix': 'mobilenet-ssd', 'data_dir': 'Y:/pascalvoc', 'mean_rgb': [127.5, 127.5, 127.5], 'file_list': 'train_person.txt', 'eval_file_list': 'eval_person.txt', 'label_list': 'label_list_person', 'mode': 'train', 'num_epochs': 400, 'train_batch_size': 32, 'use_gpu': True, 'apply_distort': True, 'apply_expand': True, 'apply_corp': True, 'image_distort_strategy': {'expand_prob': 0.5, 'expand_max_ratio': 4, 'hue_prob': 0.5, 'hue_delta': 18, 'contrast_prob': 0.5, 'contrast_delta': 0.5, 'saturation_prob': 0.5, 'saturation_delta': 0.5, 'brightness_prob': 0.5, 'brightness_delta': 0.125}, 'rsm_strategy': {'learning_rate': 0.001, 'lr_epochs': [40, 60, 80, 100], 'lr_decay': [1, 0.5, 0.25, 0.1, 0.01]}, 'momentum_strategy': {'learning_rate': 0.1, 'decay_steps': 128, 'decay_rate': 0.8}, 'early_stop': {'sample_frequency': 50, 'successive_limit': 3, 'min_loss': 1.28, 'min_curr_map': 0.86}}
W0929 14:30:00.702307 44328 device_context.cc:259] Please NOTE: device: 0, CUDA Capability: 61, Driver API Version: 10.1, Runtime API Version: 10.0
W0929 14:30:00.716289 44328 device_context.cc:267] device: 0, cuDNN Version: 7.5.
load param from retrain model
current round: 0, start read image
round 1/400, Batch 10/Unknown, loss 4.078320503234863, each batch of time 2.67 sec
round 1/400, Batch 20/Unknown, loss 5.063786506652832, each batch of time 2.23 sec
round 1/400, Batch 30/Unknown, loss 3.9244742393493652, each batch of time 2.25 sec
round 1/400, Batch 40/Unknown, loss 4.410467624664307, each batch of time 3.01 sec
round 1/400, Batch 50/Unknown, loss 4.754312992095947, each batch of time 2.55 sec
50 batch train, cur_map:0.11088856346905232 accum_map_v:0.12276104986667633 loss:4.754312992095947
round 1/400, Batch 60/Unknown, loss 4.3199462890625, each batch of time 2.16 sec
round 1/400, Batch 70/Unknown, loss 4.595076560974121, each batch of time 2.23 sec
round 1/400, Batch 80/Unknown, loss 4.2540178298950195, each batch of time 2.18 sec
round 1/400, Batch 90/Unknown, loss 4.372370719909668, each batch of time 2.41 sec
round 1/400, Batch 100/Unknown, loss 4.494147777557373, each batch of time 2.81 sec
Temp save 100 batch train result
100 batch train, cur_map:0.1763677131384611 accum_map_v:0.12373258210718632 loss:4.494147777557373
round 1/400, Batch 110/Unknown, loss 5.2544050216674805, each batch of time 2.73 sec
round 1/400, Batch 120/Unknown, loss 4.290109157562256, each batch of time 2.75 sec
round 1/400, Batch 130/Unknown, loss 4.545482635498047, each batch of time 2.72 sec
round 1/400, Batch 140/Unknown, loss 4.752466201782227, each batch of time 2.22 sec
round 1/400, Batch 150/Unknown, loss 4.584624290466309, each batch of time 2.87 sec
150 batch train, cur_map:0.15012003146111966 accum_map_v:0.12979692816734315 loss:4.584624290466309
round 1/400, Batch 160/Unknown, loss 3.9851269721984863, each batch of time 2.58 sec
round 1/400, Batch 170/Unknown, loss 4.32489013671875, each batch of time 2.54 sec
round 1/400, Batch 180/Unknown, loss 5.069633483886719, each batch of time 1.86 sec
round 1/400, Batch 190/Unknown, loss 4.214498043060303, each batch of time 2.59 sec
round 1/400, Batch 200/Unknown, loss 4.036450386047363, each batch of time 2.30 sec
Temp save 200 batch train result
200 batch train, cur_map:0.17075356356799604 accum_map_v:0.13186682462692262 loss:4.036450386047363
current round: 1, start read image
round 2/400, Batch 10/200, loss 3.9505205154418945, each batch of time 2.21 sec     
round 2/400, Batch 20/200, loss 4.849532127380371, each batch of time 2.40 sec      
round 2/400, Batch 30/200, loss 4.197256088256836, each batch of time 2.22 sec      
round 2/400, Batch 40/200, loss 5.1339311599731445, each batch of time 2.46 sec    
round 2/400, Batch 50/200, loss 4.403460502624512, each batch of time 2.78 sec      
250 batch train, cur_map:0.18925262838602067 accum_map_v:0.13546461760997772 loss:4.403460502624512
round 2/400, Batch 60/200, loss 4.496517181396484, each batch of time 3.69 sec     
round 2/400, Batch 70/200, loss 4.610703468322754, each batch of time 3.66 sec      
round 2/400, Batch 80/200, loss 4.549267292022705, each batch of time 3.18 sec     
round 2/400, Batch 90/200, loss 4.494808673858643, each batch of time 3.25 sec      
train batch 3/10, this time 2.43+0.35=2.78 sec, remaining time: 2 days 14 hours 19 mins 30 sec

**Comment by Exception-star** [comment]
我和你遇到了同样的问题，能跑一部分epoch,然后程序挂了，
<img width="543" alt="图片" src="https://user-images.githubusercontent.com/53971568/65828222-0e707400-e2cb-11e9-9f74-e979daff2d6b.png">就像这样，我不清楚问什么只能执行一些epoch，规避的方法是在原来的基础上调低batchsize的大小
